crates/matrix-sdk/src/authentication/oauth/qrcode/reciprocate/mod.rs

```diff
@@ -0,0 +1,222 @@
+    else {
+        return Err(Error::Unknown(
+            "Receiving unexpected message when expecting LoginProtocol".to_owned(),
+        ));
+    };
+
+    // We verify the selected protocol.
+    // -- MSC4108 OAuth 2.0 login step 4
+    if protocol != LoginProtocolType::DeviceAuthorizationGrant {
+        channel
+            .send_json(QrAuthMessage::LoginFailure {
+                reason: LoginFailureReason::UnsupportedProtocol,
+                homeserver: None,
+            })
+            .await?;
+        return Err(Error::UnsupportedProtocol(protocol));
+    }
+
+    // We check that the device ID is still available.
+    // -- MSC4108 OAuth 2.0 login step 4 continued
+    let request = device::get_device::v3::Request::new(device_id.to_base64().into());
+    if client.send(request).await.is_ok() {
+        channel
+            .send_json(QrAuthMessage::LoginFailure {
+                reason: LoginFailureReason::DeviceAlreadyExists,
+                homeserver: None,
+            })
+            .await?;
+        return Err(Error::DeviceIDAlreadyInUse);
+    }
+
+    // We emit an update so that the caller can open the verification URI in a
+    // system browser to consent to the login.
+    // -- MSC4108 OAuth 2.0 login step 4 continued
+    let verification_uri = Url::parse(
+        device_authorization_grant
+            .verification_uri_complete
+            .map(VerificationUriComplete::into_secret)
+            .unwrap_or(device_authorization_grant.verification_uri.to_string())
+            .as_str(),
+    )
+    .map_err(|_| Error::UnableToCreateDevice)?;
+    state.set(ReciprocateProgress::WaitingForAuth { verification_uri }.into());
+
+    // We send the new device the m.login.protocol_accepted message to let it know
+    // that the consent process is in progress.
+    // -- MSC4108 OAuth 2.0 login step 4 continued
+    let message = QrAuthMessage::LoginProtocolAccepted;
+    channel.send_json(&message).await?;
+
+    // The new device displays the user code it received from the authorization
+    // server and starts polling for an access token. In parallel, the user
+    // consents to the new login in the browser on this device, while verifying
+    // the user code displayed on the other device. -- MSC4108 OAuth 2.0 login
+    // steps 5 & 6
+
+    // We wait for the new device to send us the m.login.success message
+    let message: QrAuthMessage = channel.receive_json().await?;
+    let QrAuthMessage::LoginSuccess = message else {
+        return Err(Error::Unknown(
+            "Receiving unexpected message when expecting LoginSuccess".to_owned(),
+        ));
+    };
+
+    // We check that the new device was created successfully allowing for up to 10
+    // seconds of delay. -- MSC4108 Secret sharing and device verification step
+    // 1
+    let request = device::get_device::v3::Request::new(device_id.to_base64().into());
+    let deadline = Instant::now() + Duration::from_secs(10);
```

Comment:
```suggestion
    let deadline = Instant::now() + Duration::from_secs(10);

```

Reply:
Have applied this as part of 93ec5995219c27d824a64178ac35fdda1d3c1708.